src/main/groovy/io/sentry/android/gradle/SentryPlugin.groovy

```diff
@@ -274,60 +243,48 @@ class SentryPlugin implements Plugin<Project> {
-                                debugMetaPropPath,
-                                mappingFile
-                        ]
-
-                        if (!extension.autoUpload) {
-                            args << "--no-upload"
+                    if (VersionNumber.parse(Version.ANDROID_GRADLE_PLUGIN_VERSION) >= new VersionNumber(3, 3, 0, null)) {
+                        variant.mergeAssetsProvider.configure {
+                            dependsOn(generateUuidTask)
                         }
+                    } else {
+                        //noinspection GrDeprecatedAPIUsage
+                        variant.mergeAssets.dependsOn(generateUuidTask)
+                    }
 
+                    // create a task that persists our proguard uuid as android asset
+                    def uploadSentryProguardMappingsTask = project.tasks.create(
+                            name: "uploadSentryProguardMappings${variant.name.capitalize()}${variantOutput.name.capitalize()}",
+                            type: SentryUploadProguardMappingsTask) {
+                        dependsOn(generateUuidTask)
+                        workingDir project.rootDir
+                        getCliExecutable().set(cli)
+                        getSentryProperties().set(project.file(getPropsString(project, variant)))
+                        mappingsUuid.set(generateUuidTask.outputUuid)
+                        getMappingsFile().set(mappingFile)
+                        getAutoUpload().set(extension.autoUpload)
                         def buildTypeProperties = variant.buildType.ext
                         if (buildTypeProperties.has(SENTRY_ORG_PARAMETER)) {
-                            args.add("--org")
-                            args.add(buildTypeProperties.get(SENTRY_ORG_PARAMETER).toString())
+                            getSentryOrganization().set(buildTypeProperties.get(SENTRY_ORG_PARAMETER).toString())
                         }
                         if (buildTypeProperties.has(SENTRY_PROJECT_PARAMETER)) {
-                            args.add("--project")
-                            args.add(buildTypeProperties.get(SENTRY_PROJECT_PARAMETER).toString())
-                        }
-
-                        project.logger.info("cli args: ${args.toString()}")
-
-                        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-                            commandLine("cmd", "/c", *args)
-                        } else {
-                            commandLine(*args)
+                            getSentryProject().set(buildTypeProperties.get(SENTRY_PROJECT_PARAMETER).toString())
                         }
-
-                        project.logger.info("args executed.")
-
-                        enabled true
                     }
 
+                    variant.register(uploadSentryProguardMappingsTask)
+                    project.android.sourceSets[variant.name].assets.srcDir(generateUuidTask.outputDirectory)
```

Comment:
I guess we can ignore https://github.com/getsentry/sentry-android-gradle-plugin/pull/41/files#r577592017 because this outputs exactly to the root assets folder, no `sentry`

Reply:
I'm not sure what you mean. We are using the `outputDirectory` directory from the `generateUuidTask` task which is set to `project.file("build/generated/assets/sentry/${variant.name}")` so they will always be the same.

This is how new plugins and tasks should be written, using properties. By using properties you ensure that any changes propagate to all dependents.

Reply:
I meant that we use the `project.android.sourceSets[variant.name].assets` directory (no `/sentry`), but I asked before about the `/sentry` dir cus I've written that comment before this one, so just ignore it, all good :)

Reply:
Ah, I see. So `AndroidSourceSet.assets` is is not a directory but rather an `AndroidSourceDirectorySet` which is like a collection of directories so I'm rather adding the output directory to the source set.

In general when working with Gradle you don't want to add files to the APK/JAR/whatever by just adding files to the directory that some other task reads (such as the merged assets directory). The right way to do it is to add it to the source set and make the merge task depend on the task that generates the files.

Reply:
that makes total sense, thanks